CMakeLists.txt

```diff
@@ -0,0 +1,108 @@
+cmake_minimum_required(VERSION 3.16)
+# TODO: Get version of wgpu-native and set that up correctly here.
+project(wgpu_native LANGUAGES C)
+
+set(WGPU_NATIVE_USER_CARGO_BUILD_OPTIONS "" CACHE STRING "Additional cargo flags (such as --features) to apply to the build command")
+option(BUILD_SHARED_LIBS "Build using shared libraries" OFF)
+option(WGPU_NATIVE_ALWAYS_BUILD "If cmake should always invoke cargo to build wgpu_native" ON)
+
+# TODO: What to do about RelWithDebInfo?
+if(CMAKE_BUILD_TYPE STREQUAL "Release")
+    set(WGPU_NATIVE_BUILD_TYPE_FLAG "--release")
+    set(WGPU_NATIVE_BUILD_TYPE "release")
+else()
+    set(WGPU_NATIVE_BUILD_TYPE "debug")
+endif()
+
+if(DEFINED ENV{CARGO_BUILD_TARGET})
+    set(WGPU_NATIVE_LIB_DIR ${CMAKE_CURRENT_SOURCE_DIR}/target/$ENV{CARGO_BUILD_TARGET}/${WGPU_NATIVE_BUILD_TYPE})
+else()
+    set(WGPU_NATIVE_LIB_DIR ${CMAKE_CURRENT_SOURCE_DIR}/target/${WGPU_NATIVE_BUILD_TYPE})
+endif()
+
+if(ANDROID)
+    if(ANDROID_ABI STREQUAL "armeabi-v7a")
+        set(ANDROID_TARGET "armv7-linux-androideabi")
+        set(ANDROID_ARCH_SHORT "arm")
+    elseif(ANDROID_ABI STREQUAL "arm64-v8a")
+        set(ANDROID_TARGET "aarch64-linux-android")
+        set(ANDROID_ARCH_SHORT "aarch64")
+    elseif(ANDROID_ABI STREQUAL "x86")
+        set(ANDROID_TARGET "i686-linux-android")
+        set(ANDROID_ARCH_SHORT "i386")
+    elseif(ANDROID_ABI STREQUAL "x86_64")
+        set(ANDROID_TARGET "x86_64-linux-android")
+        set(ANDROID_ARCH_SHORT "x86_64")
+    endif()
+
+    set(WGPU_NATIVE_BUILD_TARGET "--target=${ANDROID_TARGET}")
+    # TODO: Is this needed given CARGO_BUILD_TARGET above?
+    set(WGPU_NATIVE_LIB_DIR ${CMAKE_CURRENT_SOURCE_DIR}/target/${ANDROID_TARGET}/${WGPU_NATIVE_BUILD_TYPE})
+
+    if(BUILD_SHARED_LIBS)
+        # TODO: Is this actually right?
+        message(FATAL_ERROR "Wasmtime cannot be built with BUILD_SHARED_LIBS on Android")
+    endif()
+endif()
+
+# TODO: Add mapping for iOS / etc builds to the right rust targets as well.
+
+if(BUILD_SHARED_LIBS)
+    if(WIN32)
+        # TODO: Is this actually right?
+        set(WGPU_NATIVE_BUILD_PRODUCT ${WGPU_NATIVE_LIB_DIR}/${CMAKE_SHARED_LIBRARY_PREFIX}wgpu_native.dll.lib)
+    else()
+        set(WGPU_NATIVE_BUILD_PRODUCT ${WGPU_NATIVE_LIB_DIR}/${CMAKE_SHARED_LIBRARY_PREFIX}wgpu_native${CMAKE_SHARED_LIBRARY_SUFFIX})
+    endif()
+else()
+    set(WGPU_NATIVE_BUILD_PRODUCT ${WGPU_NATIVE_LIB_DIR}/${CMAKE_STATIC_LIBRARY_PREFIX}wgpu_native${CMAKE_STATIC_LIBRARY_SUFFIX})
+endif()
+
+if (WIN32)
+    set(OS_LIBRARIES d3dcompiler ws2_32 userenv bcrypt ntdll opengl32)
+elseif(UNIX AND NOT APPLE)
+    set(OS_LIBRARIES "-lm -ldl")
+elseif(APPLE)
+    set(OS_LIBRARIES "-framework CoreFoundation -framework QuartzCore -framework Metal")
+endif()
+
+find_program(WGPU_NATIVE_CARGO_BINARY cargo)
+
+# TODO: Use Corrosion instead?
```

Comment:
I checked out Corrosion and it seems to provide everything we need. Let's try using that.
https://corrosion-rs.github.io/corrosion/usage.html